### PR TITLE
Retry some http replication failures

### PR DESCRIPTION
**Pull request by Fizzadar**
Closes: #12178

Signed off by Nick Mills-Barrett <nick@beeper.com>

### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#changelog).
* [x] Pull request includes a [sign off](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off)
* [x] [Code style](https://matrix-org.github.io/synapse/latest/code_style.html) is correct
  (run the [linters](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#run-the-linters))

### CODE REVIEW

File: synapse/replication/http/_base.py [hunk 1]
```diff
@@ -94,6 +95,8 @@ class ReplicationEndpoint(metaclass=abc.ABCMeta):
     METHOD = "POST"
     CACHE = True
     RETRY_ON_TIMEOUT = True
+    RETRY_ON_CONNECT_ERROR = True
+    RETRY_ON_CONNECT_ERROR_ATTEMPTS = 5  # =63s (2^6-1)
```
Comment (richvdh) [inline synapse/replication/http/_base.py:99]: please could you add a couple of words to the docstring above, where these attributes are documented?

File: synapse/replication/http/_base.py [hunk 2]
```diff
@@ -236,30 +239,44 @@ async def send_request(*, instance_name: str = "master", **kwargs: Any) -> Any:
...
+                            await clock.sleep(1)
+                        except (ConnectError, DNSLookupError):
+                            if not cls.RETRY_ON_CONNECT_ERROR:
+                                raise
+                            if attempts > cls.RETRY_ON_CONNECT_ERROR_ATTEMPTS:
+                                raise
+
+                            logger.warning(
+                                "%s request connection failed; retrying", cls.NAME
+                            )
```
Comment (richvdh) [inline synapse/replication/http/_base.py:276]: we should probably log the text of the failure?

```suggestion
                            logger.warning(
                                "%s request connection failed; retrying: %r", cls.NAME, e
                            )
```

(together with a change to the `except` line)

**Review by richvdh — CHANGES_REQUESTED** [review]
Thanks for this @Fizzadar, looks very sensible!

A couple of requests for quick tweaks, if you don't mind

File: synapse/replication/http/_base.py (hunk 1 above)
Comment (Fizzadar) [reply synapse/replication/http/_base.py:103]: https://github.com/matrix-org/synapse/pull/12182/commits/bde45663d04bc04902654abe457faaabb97ac136

File: synapse/replication/http/_base.py (hunk 2 above)
Comment (Fizzadar) [reply synapse/replication/http/_base.py]: https://github.com/matrix-org/synapse/pull/12182/commits/3cc2a26c7b21914cd7d58bf6e32999fe5575a6d3

**Review by richvdh — APPROVED** [review]
lgtm!

**Pull request merged**